feat(api, shared-data): Add `locatingFeatureAsParent/Child` to valid entity types

mjhuff opened the pull request:
Works toward [EXEC-77](https://opentrons.atlassian.net/browse/EXEC-77)

<!--
Thanks for taking the time to open a Pull Request (PR)! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

GitHub provides robust markdown to format your PR. Links, diagrams, pictures, and videos along with text formatting make it possible to create a rich and informative PR. For more information on GitHub markdown, see:

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

Adds `locatingFeaturesAsParent` and `locatingFeaturesAsChild` properties to the various entity types that may be involved in a stackup where relevant. This PR is best viewed commit-by-commit.

Notably, 686e4009693a80894da0e873b08e158df78a0f76 adds support for `locatingFeaturesAsParent` to all addressable areas. It sounds like we have to make this a required property even for the module addressable areas, because we can't rely on a child labware always resting on a parent module-based location. Down the road, this would imply that we will have duplicated `locatingFeaturesAsParent` data between modules and addressable areas, but I suppose there's no way around this.

<!--
Describe your PR at a high level. State acceptance criteria and how this PR fits into other work. Link issues, PRs, and other relevant resources.
-->

## Test Plan and Hands on Testing
- Covered by passing CI. There are no functional changes.
- Snapshot failures all seem expected given the PR changes.
<!--
Describe your testing of the PR. Emphasize testing not reflected in the code. Attach protocols, logs, screenshots and any other assets that support your testing.
-->

## Changelog
- Add locating feature properties to module defs, labware defs, addressable areas, and deck slots. 
<!--
List changes introduced by this PR considering future developers and the end user. Give careful thought and clear documentation to breaking changes.
-->

## Review requests
- It will make accessing `locatingFeatureAsParent` much cleaner in our utils if the addressable area property is `locatingFeatureAsParent`, camel-cased, which is consistent with the camel-cased `locatingFeaturesAsParent` present in labware defs. However, this does break the current `AddressableArea` convention of snake-casing properties. Are we ok with this or not?
- Does the above logic concerning requiring all addressable areas to specify a `locatingFeatureAsParent` seem valid?
<!--
- What do you need from reviewers to feel confident this PR is ready to merge?
- Ask questions.
-->

## Risk assessment
none
<!--
- Indicate the level of attention this PR needs.
- Provide context to guide reviewers.
- Discuss trade-offs, coupling, and side effects.
- Look for the possibility, even if you think it's small, that your change may affect some other part of the system.
  - For instance, changing return tip behavior may also change the behavior of labware calibration.
- How do your unit tests and on hands on testing mitigate this PR's risks and the risk of future regressions?
- Especially in high risk PRs, explain how you know your testing is enough.
-->


[EXEC-77]: https://opentrons.atlassian.net/browse/EXEC-77?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mjhuff committed:
feat(api, shared-data): add locatingFeaturesAsParent support for modules.

mjhuff committed:
feat(api): add locatingFeaturesAsParent/Child support for labware.

mjhuff committed:
refactor(api): add labwareInterfaceX/YDimension to module def type

mjhuff committed:
feat(api): add locatingFeaturesAsParent to addressable areas

mjhuff committed:
feat(api): add to_point method on moduleoffset

mjhuff committed:
feat(api): locatingFeaturesAsParent for slotdefv3

sfoster1 reviewed (approved):
I think it's a good idea to make everything have the feature, yes. I think it is annoying to have to special case code for the underscores but it's worth it to keep the attribute names consistent. I wish the other stuff was snake case too 😢

mjhuff committed:
🐍

sfoster1 reviewed (approved):
🐍 🐍 🐍 🐍 🐍

mjhuff commented:
The underlying problem we now come up against is that `loadModule` commands contain the module definition which means we have to do database migration...or not: [EXEC-212](https://opentrons.atlassian.net/browse/EXEC-212).

After consulting with @SyntaxColoring , the plan is to make  `locatingFeaturesAsParent` as optional for now on for module defs, then do EXEC-212 immediately, and then make this field required again. So no snapshot changes for now.

mjhuff committed:
optional lfs for module defs...for now

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal locating-feature-support snapshots (#18635)

This PR was requested on the PR
https://github.com/Opentrons/opentrons/pull/18628

Co-authored-by: mjhuff <64858653+mjhuff@users.noreply.github.com>
